src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +673,482 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+            if (!consensus) {
+                errored_partitions.emplace_back(i, req.ktp().get_partition());
+                continue;
+            }
+            last_visible_indexes[i] = consensus->last_visible_index();
+        }
+
+        // A read_result needs to be returned for every partition. Hence,
+        // the function can't return before calling
+        // `fetch_ntps_in_parallel`.
+        std::vector<read_result> results = co_await fetch_ntps_in_parallel(
+          _ctx.mgr,
+          _ctx.srv.get_replica_selector(),
+          std::move(requests),
+          _ctx.foreign_read,
+          _ctx.deadline,
+          _ctx.bytes_left,
+          _ctx.srv.memory(),
+          _ctx.srv.memory_fetch_sem());
+
+        // If we weren't able to read the last_visible_index for a partition
+        // before calling `fetch_ntps_in_parallel` then we need to
+        // return with an error for that partition.
+        for (auto [i, partition] : errored_partitions) {
+            results[i] = read_result(error_code::not_leader_for_partition);
+            results[i].partition = partition;
+        }
+
+        for (const auto& r : results) {
+            total_size += r.data_size_bytes();
+            if (r.error != error_code::none) {
+                has_error = true;
+            }
+        }
+
+        co_return query_results{
+          .last_visible_indexes = std::move(last_visible_indexes),
+          .has_error = has_error,
+          .results = std::move(results),
+          .total_size = total_size,
+        };
+    }
+
+    std::optional<size_t> register_waiters(const auto& request_indexes) {
+        for (size_t i : request_indexes) {
+            auto part = _ctx.mgr.get(_ctx.requests[i].ktp());
+            // If the partition can't be found then it's since been moved
+            if (!part) {
+                return {i};
+            }
+
+            auto consensus = part->raft();
+            if (!consensus) {
+                return {i};
+            }
+
+            auto offset = model::next_offset(_last_visible_indexes[i]);
+            auto waiter = consensus->visible_offset_monitor()
+                            .wait(offset, model::no_timeout, _as)
+                            // All exceptions are ignored here as this is only
+                            // used to signal the worker that another attempt to
+                            // read the partition should be made.
+                            .handle_exception([](const std::exception_ptr&) {});
+
+            ssx::spawn_with_gate(
+              _waiter_gate, [this, w = std::move(waiter), i]() mutable {
+                  return w.finally([this, i] {
+                      _request_indexes.push_back(i);
+                      _has_completed_waiters.signal();
+                  });
+              });
+        }
+
+        return {};
+    }
+
+    ss::future<worker_result> do_run() {
+        bool first_run{true};
+        bool has_error{false};
```

Comment:
This not used in this outer scope, only in the inner scope where it can be directly initalized.

Reply:
Nice catch, removed it from the outer scope.